[iOS] Scrolling after element focus does not avoid input views when OOP keyboard is enabled

whsieh opened the pull request:
#### beacee6a1c37d359baa414a90aa73da91247cbd6
<pre>
[iOS] Scrolling after element focus does not avoid input views when OOP keyboard is enabled
<a href="https://bugs.webkit.org/show_bug.cgi?id=249159">https://bugs.webkit.org/show_bug.cgi?id=249159</a>
rdar://88057475

Reviewed by Aditya Keerthi.

Scrolling to reveal the focused element is currently broken when OOP keyboard is enabled. This is
because OOP keyboard causes `UIKeyboardWillShowNotification` to be delivered to the app
asynchronously (after communicating with the input UI process), rather than dispatched immediately
under the call to `-reloadInputViews`. WebKit currently depends on the latter, since we expect the
keyboard&apos;s final geometry (after animation) to be set by the time the post-layout editor state
arrives in the UI process and we reveal the focused element using `-_zoomToRevealFocusedElement`.
However, since this geometry is sent through the &quot;KeyboardWillShow&quot; notification which (normally)
arrives after the editor state, we end up proceeding with `-_zoomToRevealFocusedElement` without
knowing the final keyboard geometry, so we end up only scrolling to keep the field centered within
the web view (ignoring the obscuring bounds of the software keyboard). In some cases (e.g. when
focusing an input field in some system views), this can cause focused input fields to be completely
obscured by the keyboard.

This is somewhat tricky to fix, since it&apos;s not guaranteed that the &quot;KeyboardWillShow&quot; notification
will always arrive after the editor state or vice versa. To complicate things further, in the Mail
compose case (i.e. when the web view&apos;s scroller has `-firstResponderKeyboardAvoidanceEnabled` set to
`NO`), we&apos;ll actually zoom to reveal the focused element only after `UIKeyboardDidShowNotification`
instead, since the keyboard geometry isn&apos;t available before the animation has finished (see
<a href="https://commits.webkit.org/246159@main">https://commits.webkit.org/246159@main</a> for more context). And of course, this fix also needs to
continue working with OOP keyboard disabled, where the &quot;KeyboardWillShow&quot; notification is dispatched
before we&apos;ve even begun waiting for the next editor state.

To handle all of the above scenarios, we add a helper class that encapsulates logic for deferring
calls to `-_zoomToRevealFocusedElement` after an element is focused. This helper class,
`WebKit::RevealFocusedElementDeferrer`, maintains a set of flags that enumerate the three distinct
reasons why we could be deferring this call: &quot;waiting for an editor state&quot;, &quot;waiting for a
`KeyboardWillShow` notification&quot;, and &quot;waiting for a `KeyboardDidShow` notification&quot;. As
`WKContentView` receives keyboard notification from UIKit or receives an editor state update from
the web process, it removes the appropriate deferral reasons from this helper object. Finally, when
there are no longer any deferral reasons, we call into `-_zoomToRevealFocusedElement` and clear out
the helper.

Test: fast/forms/ios/scroll-after-tapping-on-input-with-software-keyboard.html

* LayoutTests/fast/forms/ios/scroll-after-tapping-on-input-with-software-keyboard-expected.txt: Added.
* LayoutTests/fast/forms/ios/scroll-after-tapping-on-input-with-software-keyboard.html: Added.

Add a new layout test to exercise this scenario. All existing layout tests are actually insufficient
to cover the changes in this fix, since a test needs to:

-   Show the software keyboard (by default, layout tests run with the hardware keyboard connected).
-   Check that we scrolled the focused input into (roughly) the center of the visible viewport, not
    just onto the screen.

* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/ios/RevealFocusedElementDeferrer.h: Added.
* Source/WebKit/UIProcess/ios/RevealFocusedElementDeferrer.mm: Added.
(WebKit::RevealFocusedElementDeferrer::RevealFocusedElementDeferrer):
(WebKit::RevealFocusedElementDeferrer::create):
(WebKit::RevealFocusedElementDeferrer::fulfill):

Add the new helper class. We protect the deferrer here since it expects its client (`WKContentView`)
to clear any references to itself when `-_zoomToRevealFocusedElement` is invoked.

* Source/WebKit/UIProcess/ios/WKContentViewInteraction.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView cleanUpInteraction]):
(-[WKContentView _keyboardWillShow]):
(-[WKContentView _keyboardDidShow]):
(-[WKContentView _zoomToRevealFocusedElement]):
(-[WKContentView resetShouldZoomToFocusRectAfterShowingKeyboard]):
(-[WKContentView _elementDidFocus:userIsInteracting:blurPreviousNode:activityStateChanges:userObject:]):

Either call `-_zoomToRevealFocusedElement` immediately, or create a `RevealFocusedElementDeferrer`,
passing in deferral reasons as appropriate. See comments above for more details.

(-[WKContentView _elementDidBlur]):
(-[WKContentView _didUpdateEditorState]):
(-[WKContentView _zoomToFocusRectAfterShowingKeyboardIfNeeded]): Deleted.
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/257785@main">https://commits.webkit.org/257785@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1df010032f30bbf010af5b1c4cd378d91c4eda53

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/99909 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/9081 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/32991 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/109262 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/169500 "Built successfully and passed tests") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/9962 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/86375 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/92359 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/107165 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/105676 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/7516 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/90804 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/34245 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/89422 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/22194 "Passed tests") | [⏳ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/API-Tests-GTK-EWS "Waiting to run tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/2882 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/23708 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/2838 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/46080 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5347 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/8971 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/43181 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/4690 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
```diff
@@ -350,6 +350,28 @@ - (void)setImageAnalysis:(CocoaImageAnalysis *)analysis;
     return stream;
 }
 
+RevealFocusedElementDeferrer::RevealFocusedElementDeferrer(WKContentView *view, OptionSet<RevealFocusedElementDeferralReason> reasons)
```
We should consider moving this into it's own file, given how bloated `WKContentViewInteraction.mm` is 😅

whsieh replied:
Good call — will move to a `RevealFocusedElementDeferrer.{h|mm}`.

Thanks for the review!

whsieh committed:
[iOS] Scrolling after element focus does not avoid input views when OOP keyboard is enabled
https://bugs.webkit.org/show_bug.cgi?id=249159
rdar://88057475

Reviewed by Aditya Keerthi.

Scrolling to reveal the focused element is currently broken when OOP keyboard is enabled. This is
because OOP keyboard causes `UIKeyboardWillShowNotification` to be delivered to the app
asynchronously (after communicating with the input UI process), rather than dispatched immediately
under the call to `-reloadInputViews`. WebKit currently depends on the latter, since we expect the
keyboard's final geometry (after animation) to be set by the time the post-layout editor state
arrives in the UI process and we reveal the focused element using `-_zoomToRevealFocusedElement`.
However, since this geometry is sent through the "KeyboardWillShow" notification which (normally)
arrives after the editor state, we end up proceeding with `-_zoomToRevealFocusedElement` without
knowing the final keyboard geometry, so we end up only scrolling to keep the field centered within
the web view (ignoring the obscuring bounds of the software keyboard). In some cases (e.g. when
focusing an input field in some system views), this can cause focused input fields to be completely
obscured by the keyboard.

This is somewhat tricky to fix, since it's not guaranteed that the "KeyboardWillShow" notification
will always arrive after the editor state or vice versa. To complicate things further, in the Mail
compose case (i.e. when the web view's scroller has `-firstResponderKeyboardAvoidanceEnabled` set to
`NO`), we'll actually zoom to reveal the focused element only after `UIKeyboardDidShowNotification`
instead, since the keyboard geometry isn't available before the animation has finished (see
https://commits.webkit.org/246159@main for more context). And of course, this fix also needs to
continue working with OOP keyboard disabled, where the "KeyboardWillShow" notification is dispatched
before we've even begun waiting for the next editor state.

To handle all of the above scenarios, we add a helper class that encapsulates logic for deferring
calls to `-_zoomToRevealFocusedElement` after an element is focused. This helper class,
`WebKit::RevealFocusedElementDeferrer`, maintains a set of flags that enumerate the three distinct
reasons why we could be deferring this call: "waiting for an editor state", "waiting for a
`KeyboardWillShow` notification", and "waiting for a `KeyboardDidShow` notification". As
`WKContentView` receives keyboard notification from UIKit or receives an editor state update from
the web process, it removes the appropriate deferral reasons from this helper object. Finally, when
there are no longer any deferral reasons, we call into `-_zoomToRevealFocusedElement` and clear out
the helper.

Test: fast/forms/ios/scroll-after-tapping-on-input-with-software-keyboard.html

* LayoutTests/fast/forms/ios/scroll-after-tapping-on-input-with-software-keyboard-expected.txt: Added.
* LayoutTests/fast/forms/ios/scroll-after-tapping-on-input-with-software-keyboard.html: Added.

Add a new layout test to exercise this scenario. All existing layout tests are actually insufficient
to cover the changes in this fix, since a test needs to:

-   Show the software keyboard (by default, layout tests run with the hardware keyboard connected).
-   Check that we scrolled the focused input into (roughly) the center of the visible viewport, not
    just onto the screen.

* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/ios/RevealFocusedElementDeferrer.h: Added.
* Source/WebKit/UIProcess/ios/RevealFocusedElementDeferrer.mm: Added.
(WebKit::RevealFocusedElementDeferrer::RevealFocusedElementDeferrer):
(WebKit::RevealFocusedElementDeferrer::create):
(WebKit::RevealFocusedElementDeferrer::fulfill):

Add the new helper class. We protect the deferrer here since it expects its client (`WKContentView`)
to clear any references to itself when `-_zoomToRevealFocusedElement` is invoked.

* Source/WebKit/UIProcess/ios/WKContentViewInteraction.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView cleanUpInteraction]):
(-[WKContentView _keyboardWillShow]):
(-[WKContentView _keyboardDidShow]):
(-[WKContentView _zoomToRevealFocusedElement]):
(-[WKContentView resetShouldZoomToFocusRectAfterShowingKeyboard]):
(-[WKContentView _elementDidFocus:userIsInteracting:blurPreviousNode:activityStateChanges:userObject:]):

Either call `-_zoomToRevealFocusedElement` immediately, or create a `RevealFocusedElementDeferrer`,
passing in deferral reasons as appropriate. See comments above for more details.

(-[WKContentView _elementDidBlur]):
(-[WKContentView _didUpdateEditorState]):
(-[WKContentView _zoomToFocusRectAfterShowingKeyboardIfNeeded]): Deleted.
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: https://commits.webkit.org/257785@main
